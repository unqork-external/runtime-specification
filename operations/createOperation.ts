import type { OperationTypes } from './enums/operation-types.enum'
import { OperationTypeMap } from './typemaps/operation.typemap'
import type { CustomOperationTypeMap } from '../../operations'
import type { InputRef } from '../inputs/inputRef'

/**
 * Given an OperationType, Typescript will complete the options list.
 *
 * In return, the caller receives Typescript constructed class.
 *
 * This includes preset defaults.
 *
 * In essence, this function ensures that all keys are fully present, fully typed, and ready for runtime
 * validation if necessary.
 *
 * @param type
 * @param options
 * @param name name of the operation
 * @param additional
 * @param spreadOp
 * @param additionalArgs
 * @returns - The Operation Class type
 */
export const createOperation = <OpType extends OperationTypes>(
  type: OpType,
  options: InstanceType<(typeof OperationTypeMap)[OpType]>['options'],
  additionalArgs?: Partial<AdditionalArgs>,
): InstanceType<(typeof OperationTypeMap)[OpType]> => {
  const classOpType = OperationTypeMap[type]
  const op = new classOpType()
  return Object.assign(op, {
    type,
    options: {
      ...op.options,
      ...options,
    },
    ...additionalArgs,
  })
}

export const createCustomOperation = <OperationEnum extends string, OpType extends OperationEnum>(
  type: OpType,
  options: InstanceType<CustomOperationTypeMap<OperationEnum>[OpType]['model']>['options'],
  typeMap: CustomOperationTypeMap<OperationEnum>,
  additionalArgs?: Partial<AdditionalArgs>,
) => {
  const classOpType = typeMap[type].model
  const op = new classOpType()
  return Object.assign(op, {
    type,
    options: {
      ...op.options,
      ...options,
    },
    ...additionalArgs,
  })
}

interface AdditionalArgs {
  name: string
  notifyImmediately: boolean
  creatorSummary: string
}
