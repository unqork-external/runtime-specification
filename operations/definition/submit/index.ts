export * from './submit.options'
export * from './submit.model'
