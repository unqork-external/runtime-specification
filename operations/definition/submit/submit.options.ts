import { Description, Example, Required } from '@tsed/schema'

import { BaseOperationOptions } from '../../interface/operations.interface'

export class SubmitOptions extends BaseOperationOptions {
  @Required()
  @Description('Key of a target entity')
  @Example('submissionData')
  declare targetEntityKey: string
}
