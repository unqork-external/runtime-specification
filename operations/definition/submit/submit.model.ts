import { DiscriminatorValue, Required } from '@tsed/schema'

import { SubmitOptions } from './submit.options'
import { Alpha } from '../../../../decorators'
import { AutogenDisabled } from '../../decorators/autogenDisabled.decorator'
import { OperationTypes } from '../../enums/operation-types.enum'
import { Operation } from '../../interface/operations.interface'

@DiscriminatorValue(OperationTypes.SUBMIT)
@Alpha()
@AutogenDisabled()
export class SubmitOperation extends Operation<OperationTypes.SUBMIT> {
  public type = OperationTypes.SUBMIT as const

  @Required()
  public options: SubmitOptions = new SubmitOptions()
}
