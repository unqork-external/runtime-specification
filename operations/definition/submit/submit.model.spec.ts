import { SubmitOperation } from './submit.model'
import { type SubmitOptions } from './submit.options'
import { generateSchemaAndValidate } from '../../../../utilities'
import { createOperation } from '../../createOperation'
import { OperationTypes } from '../../enums/operation-types.enum'

describe('Submit Model', () => {
  let validate, schema
  beforeAll(() => {
    const response = generateSchemaAndValidate(SubmitOperation)
    validate = response.validate
    schema = response.schema
  })

  it('should match this JSON Schema', () => {
    expect(schema).toMatchSnapshot()
  })

  it('should validate submit operation with all required keys', () => {
    const op = createOperation(OperationTypes.SUBMIT, {
      targetEntityKey: 'firstName',
    })

    const isValid = validate(op)
    expect(isValid).toBeTrue()
  })

  it('should error if no targetEntityKey is provided', () => {
    const op = createOperation(OperationTypes.SUBMIT, {} as SubmitOptions)

    const isValid = validate(op)
    expect(isValid).toBeFalse()
  })
})
