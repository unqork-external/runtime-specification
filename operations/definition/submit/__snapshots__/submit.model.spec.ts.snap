// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Submit Model should match this JSON Schema 1`] = `
{
  "autogenDisabled": true,
  "definitions": {
    "SubmitOptions": {
      "properties": {
        "shouldExecute": {
          "default": true,
          "description": "
    Conditionally execute this operation.
    Can set to false to temporarily disabled when debugging.
  ",
          "type": [
            "string",
            "boolean",
          ],
        },
        "targetEntityKey": {
          "description": "Key of a target entity",
          "examples": [
            "submissionData",
          ],
          "minLength": 1,
          "type": "string",
        },
      },
      "required": [
        "targetEntityKey",
      ],
      "type": "object",
    },
  },
  "properties": {
    "creatorSummary": {
      "description": "A detailed summary of the operation",
      "type": "string",
    },
    "name": {
      "description": "Name of the operation",
      "type": "string",
    },
    "options": {
      "$ref": "#/definitions/SubmitOptions",
    },
    "type": {
      "const": "SUBMIT",
      "examples": [
        "SUBMIT",
      ],
      "type": "string",
    },
  },
  "required": [
    "options",
  ],
  "type": "object",
}
`;
